metagpt/provider/google_gemini_api.py

```diff
@@ -0,0 +1,152 @@
+                completion_tokens = int(usage.get("completion_tokens", 0))
+                self._cost_manager.update_cost(prompt_tokens, completion_tokens, self.model)
+            except Exception as e:
+                logger.error(f"google gemini updats costs failed! exp: {e}")
+
+    def get_choice_text(self, resp: GenerateContentResponse) -> str:
+        return resp.text
+
+    def get_usage(self, messages: list[dict], resp_text: str) -> dict:
+        req_text = messages[-1]["parts"][0] if messages else ""
+        prompt_resp = self.llm.count_tokens(contents={"role": "user", "parts": [{"text": req_text}]})
+        completion_resp = self.llm.count_tokens(contents={"role": "model", "parts": [{"text": resp_text}]})
+        usage = {
+            "prompt_tokens": prompt_resp.total_tokens,
+            "completion_tokens": completion_resp.total_tokens
+        }
+        return usage
+
+    async def aget_usage(self, messages: list[dict], resp_text: str) -> dict:
+        req_text = messages[-1]["parts"][0] if messages else ""
+        prompt_resp = await self.llm.count_tokens_async(contents={"role": "user", "parts": [{"text": req_text}]})
+        completion_resp = await self.llm.count_tokens_async(contents={"role": "model", "parts": [{"text": resp_text}]})
+        usage = {
+            "prompt_tokens": prompt_resp.total_tokens,
+            "completion_tokens": completion_resp.total_tokens
+        }
+        return usage
+
+    def completion(self, messages: list[dict]) -> "GenerateContentResponse":
+        resp: GenerateContentResponse = self.llm.generate_content(**self._const_kwargs(messages))
+        usage = self.get_usage(messages, resp.text)
+        self._update_costs(usage)
+        return resp
+
+    async def _achat_completion(self, messages: list[dict]) -> "AsyncGenerateContentResponse":
+        resp: AsyncGenerateContentResponse = await self.llm.generate_content_async(**self._const_kwargs(messages))
+        usage = await self.aget_usage(messages, resp.text)
+        self._update_costs(usage)
+        return resp
+
+    async def acompletion(self, messages: list[dict]) -> dict:
+        return await self._achat_completion(messages)
+
+    async def _achat_completion_stream(self, messages: list[dict]) -> str:
+        resp: AsyncGenerateContentResponse = await self.llm.generate_content_async(**self._const_kwargs(messages,
+                                                                                                        stream=True))
+        collected_content = []
+        async for chunk in resp:
+            content = chunk.text
+            print(content, end="")
+            collected_content.append(content)
+
+        full_content = "".join(collected_content)
+        usage = await self.aget_usage(messages, full_content)
+        self._update_costs(usage)
+        return full_content
+
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_fixed(1),
```

Comment:
if gemini api is similar to openai, using "wait_random_exponential" will be better